### ISSUE TITLE
Neovide Threading Issue

### ISSUE THREAD

**Issue by moncheeta** [neovide/neovide#1068]
I try to open Neovide from the terminal and it doesn't open and returns an error. Other times, it works. I'm on Arch Linux with NVIM Version 6. I am running the "without helper" version under Linux Install.
What it returns when I run "neovide":

```
[xcb] Unknown sequence number while processing queue
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
neovide: xcb_io.c:269: poll_for_event: Assertion `!xcb_xlib_threads_sequence_lost' failed.
```

**Comment by Moncheeta** [comment]
Here is my init.vim file if you need it:
set guifont=JetBrains\ Mono:h16
map <C-n> :NERDTreeToggle<CR>
call plug#begin('~/.config/nvim/autoload/plugged')
Plug 'sheerun/vim-polyglot'
Plug 'scrooloose/NERDTree'
Plug 'jiangmiao/auto-pairs'
call plug#end()

**Comment by PyGamer0** [comment]
Which version of neovide?

**Comment by Moncheeta** [comment]
I am on version 0.8.0.

**Comment by Moncheeta** [comment]
When I put my laptop to sleep and turn it on, I get:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 4, kind: Interrupted, message: "Interrupted system call" }', /home/moncheeta/.cache/yay/neovide-git/src/neovide-git/CARGO/git/checkouts/winit-cdba476763ae8f28/3b5462e/src/platform_impl/linux/x11/mod.rs:396:54
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by Moncheeta** [comment]
Does anyone know how to fix it?

**Comment by Kethku** [comment]
I would look upstream for similar errors in winit. Seems likely that this is an upstream issue and further that it may already be fixed.

Unfortunately we are using a fork that needs updated if we determine this problem has already been fixed.

**Comment by MultisampledNight** [comment]
May we know which GPU and driver are you on? I got the same thing on a pretty old laptop and haven't been able to fix it either.

**Comment by fredizzimo** [comment]
I did some research. Winit did have an issue like this before, caused by a race condition, but fixed a long time ago. https://github.com/rust-windowing/winit/issues/458.

The fix was further improved in https://github.com/rust-windowing/winit/commit/28e3c35547609ac40a472f656f51e54882a793

But it's possible that something in the new keyboard API handling has broken it again. A full backtrace would be helpful to see exactly where the problem comes from, so set the environment variable `RUST_BACKTRACE=1` before running. Unless we can figure it out, by looking at that, I guess an issue should be open in the `rust-windowing' repository, since they probably have a better chance to know what the issue is.

**Comment by asteroidalaz** [comment]
Stills it is same with RUST_BACKTRACE=1 and full
```
[xcb] Unknown sequence number while processing queue
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
neovide: xcb_io.c:269: poll_for_event: Assertion `!xcb_xlib_threads_sequence_lost' failed.
```

**Comment by Moncheeta** [comment]
I have tried it on two laptops. One is an Intel integrated graphics on an Intel i7 2nd gen. The other is a AMD A6-4455M with AMD ATI Radeon HD 7500G. So far, this has been happening on all of my devices.

**Comment by mmulli** [comment]
Hi, I'm getting the same errors on Arch Linux (Roughly 50% of the time and on version 0.8.0). Any more thoughts on this?

**Comment by fredizzimo** [comment]
By (st or Alacritty, do you mean that you get the same error when opening just the terminal too, or that you get it when running Neovide from the terminal? 

But yes, it's related to the Window Manager, specifically X and the `X keyboard extension`. that part of Neovide is handled by teh `winit` library, more specifically by our custom fork of it, including these changes https://github.com/rust-windowing/winit/pull/1932. 

As far as I know we are the only real project using that at the moment, so it's likely the bug is either in those change, or that we haven't merged everything correctly. So we need to check that our fork is up to date, and then discuss with the upstream. I think they have a better chance of tracking this down. Especially since I don't have a system to repeat the bug here.

**Comment by mmulli** [comment]
By st or Alacrity i mean that I get the error when running Neovide from either of those.

And understood regarding the fork. Hopefully it's just behind a bit and solved upstream already. See what happens I guess, have a good weekend.

**Comment by SyedAhkam** [comment]
> For those that have replicated this issue. Can you please let me know what Desktop Manager or Window Manager you are running (If you've setup something custom)

I have replicated this issue on my old laptop.

### The configuration

- Intel HD 4000
- ArchLinux
- i3 window manager (X)

Me and @Kethku recently had a discussion regarding this in our discord server. And we came to the conclusion that this still hasn't fixed upstream in winit.

@mmulli Could you try compiling the [test-xcb-bug](https://github.com/neovide/neovide/tree/test-x11-bug) branch and report?

**Comment by mmulli** [comment]
@SyedAhkam Thanks for sharing your config. I see there are a few things in common with mine.

I pulled and built the **test-x11-bug** branch and got the same error output. And also when the window titled "winit window" does show it's about 1 in 5 tries that it shows. And it's an empty window that I can see through to my desktop.

Output from my console:
```
[xcb] Unknown sequence number while processing queue
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
myneovide: xcb_io.c:269: poll_for_event: Assertion `!xcb_xlib_threads_sequence_lost' failed.
zsh: abort (core dumped)  myneovide
```

**Comment by mmulli** [comment]
Let me know if you need help testing anything and ill jump on it

**Comment by nowave7** [comment]
I can also confirm this happening quite often, mostly after laptop sleep/wake cycle, but also sometimes during normal work when neovide has not been touched for some time. The configuration is Linux Mint 20.3 running Cinnamon 5.2.7 on Intel I7-4600U and Intel Haswell ULT integrated video, the driver is i915.

**Comment by fredizzimo** [comment]
Since the `test-x11-bug` branch appears to use the non-forked version of winit, it either means that we are doing something different than other programs using it, or that there's a bug. 

Could you try some of the software listed here https://github.com/rust-windowing/winit/wiki/Software-that-uses-winit, and see if you get the same kinds of errors?

**Comment by nowave7** [comment]
> Since the `test-x11-bug` branch appears to use the non-forked version of winit, it either means that we are doing something different than other programs using it, or that there's a bug.
> 
> Could you try some of the software listed here https://github.com/rust-windowing/winit/wiki/Software-that-uses-winit, and see if you get the same kinds of errors?

Hey fredizzimo,
Yes, of course> I will try the alacritty, since the other ones are not so interesting to me. I'll see to play with it for some days and then report back here if I've encountered any issues.

**Comment by mmulli** [comment]
Just a quick heads up that I've been using Alacritty and have not had this error.  Though I have not tested putting my computer to sleep and waking it up with Alacritty running. But regardless of sleep / awake, I get the error about 50% of the time with Neovide. I'll try one of the other links too.

**Comment by nowave7** [comment]
> Just a quick heads up that I've been using Alacritty and have not had this error. Though I have not tested putting my computer to sleep and waking it up with Alacritty running. But regardless of sleep / awake, I get the error about 50% of the time with Neovide. I'll try one of the other links too.

That's good to know. I'll try with sleep/wake scenario then. What I've also noticed is random freezes on Windows, again, on laptop, and again, mostly after sleep/wake cycle. I can try Alacritty on Windows as well, and see how that behaves there...
I'll report back when I have something concrete.

**Comment by SyedAhkam** [comment]
I have some interesting updates about this pesky bug!

I was just working on a side-project of mine (on my old Linux laptop) and experienced that error for the first time in an application other than Neovide.

After a bit of tinkering, I created a minimum [repro](https://github.com/SyedAhkam/glutin-x11-bug/blob/master/src/main.rs). Feel free to give it a try.

Yes, that somehow only impacts glutin. I couldn't reproduce the issue on bare winit. (eventho glutin basically mirrors the winit api).

After a bit more tinkering, Interestingly there's some point noting:

- It crashes on i3, dwm, but not on leftwm (written in rust only)
- My repro crashes 100% of the time.
- Relogging, rebooting doesn't help. It starts crashing from the moment you start up your machine.
- Glutin's `ContextBuilder` seems to be the breaking point. Commenting that out fixes the issue and we are greeted by a sweet window.

**Comment by nowave7** [comment]
OK, here are my two cents after playing around with Alacritty and Neovide for a couple of days in parallel.
Neovide did crash on two occasions, while Alacritty hasn't. Now, I know this is a relatively small time frame to draw any meaningful conclusions, so I'll just keep doing the sleep/awake cycle with both apps running, and update once I have more concrete data

@SyedAhkam 
I'll also try out your repo and see if I can reproduce this bug on a Cinnamon running machine.

**Comment by Kethku** [comment]
I'm almost certain that this is an issue that is fixed in an upstream update to the keyboard api rework branch. So we will have to wait for a branch which contains the rework for all of the platforms before proceeding.

**Comment by mahkoh** [comment]
This is probably https://gitlab.freedesktop.org/mesa/mesa/-/issues/108

**Comment by hatemets** [comment]
Any updates on this issue? Or, alternatively, a functioning workaround? I keep having the same issue on my Arch virtual machine.

**Comment by SyedAhkam** [comment]
Hey @moncheeta 

Please provide an update. Was the issue resolved on your end?

**Comment by moncheeta** [comment]
I have forgot about this issue. So, I tried launching it again recently and It seems to work fine. It even works between sleeps and resumes on my laptop. But, please note, I'm not on Arch Linux anymore, so I don't know if it's been fixed there.

**Comment by fredizzimo** [comment]
It's quite likely that this was fixed in main for a while and now in the 0.11.0 release of Neovide. Winit was updated and completely rewrote the keyboard handling, which is where this crash happened, at least based on the filename.

But if you @SyedAhkam, still see crashes with 0.11.0, I will reopen the issue.

**Comment by SyedAhkam** [comment]
It's alright from my end actually. My old laptop (where I reproduced this earlier) is dead now and I have no way to test this out currently.

I'm quite positive it has been fixed in the 0.11.0. Let's keep this closed as is.

ps: sorry for necrobumping, just wished for this to close with a closure and not mysteriously if someone stumbles upon this in the future.
